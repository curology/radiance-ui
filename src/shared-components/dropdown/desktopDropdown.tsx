import React from 'react';
import { css, useTheme } from '@emotion/react';

import useResetFocus from '../../utils/accessibility/useResetFocus';
import { isDefined } from '../../utils/isDefined';
import { keyboardKeys } from '../../constants/keyboardKeys';
import { OffClickWrapper } from '../offClickWrapper';
import { ChevronIcon } from '../../icons';
import Style from './style';

import { OptionType, OptionValue } from './index';

interface DesktopDropdownProps<T> {
  borderRadius: string;
  closeDropdown: () => void;
  currentOption?: T;
  id?: string;
  isOpen: boolean;
  onDesktopSelectChange: (
    event: React.MouseEvent<HTMLLIElement> | React.KeyboardEvent<HTMLLIElement>,
  ) => void;
  onDropdownContainerFocus: (event: React.FocusEvent<HTMLDivElement>) => void;
  options: T[];
  optionsContainerMaxHeight: string;
  textAlign: 'left' | 'center';
  toggleDropdown: () => void;
  value?: OptionValue;
}

export const DesktopDropdown = <T extends OptionType>({
  borderRadius,
  closeDropdown,
  currentOption,
  id,
  isOpen,
  onDesktopSelectChange,
  onDropdownContainerFocus,
  options,
  optionsContainerMaxHeight,
  textAlign,
  toggleDropdown,
  value,
}: DesktopDropdownProps<T>) => {
  const theme = useTheme();
  const { initialFocus, resetFocus } = useResetFocus<HTMLDivElement>();

  const handleKeyDown = (event: React.KeyboardEvent) => {
    // This key handler allows users to open the dropdown options via the keyboard
    if (event.key === keyboardKeys.enter) {
      toggleDropdown();
    }
  };

  const handleOptionKeydown = (event: React.KeyboardEvent<HTMLLIElement>) => {
    // This allows users to select an option via the enter key
    if (event.key === keyboardKeys.enter) {
      onDesktopSelectChange(event);
      resetFocus();
    }
  };

  return (
    <OffClickWrapper
      onOffClick={closeDropdown}
      css={css`
        width: 100%;
      `}
    >
      <Style.DropdownContainer textAlign={textAlign}>
        <Style.DropdownFocusContainer
          onClick={toggleDropdown}
          onKeyDown={handleKeyDown}
          tabIndex={0}
          aria-haspopup="menu"
          role="button"
          ref={initialFocus}
          onFocus={onDropdownContainerFocus}
        >
          <div
            css={Style.dropdownInputStyle({
              borderRadius,
              shouldBeFullyRounded: !isOpen,
              textAlign,
              theme,
            })}
          >
            {currentOption?.label}
          </div>
          <Style.IconContainer>
            <ChevronIcon
              width={10}
              height={10}
              rotate={isOpen ? 90 : 0}
            />
          </Style.IconContainer>
        </Style.DropdownFocusContainer>

        <Style.DropdownOptionsContainer
          borderRadius={borderRadius}
          id={id}
          isOpen={isOpen}
          optionsContainerMaxHeight={optionsContainerMaxHeight}
          role="menu"
          aria-activedescendant={isDefined(value) ? `${value}` : undefined}
          aria-hidden={!isOpen}
        >
          {options.map((option, index) => {
            const {
              value: optionValue,
              disabled = false,
              label,
              ...rest
            } = option;

            const optionId = isDefined(optionValue)
              ? `${optionValue}`
              : `undefined-${index}`;

            return (
              <Style.DropdownOption
                key={optionId}
                value={optionValue}
                id={optionId}
                selected={value === optionValue}
                disabled={disabled}
                onClick={onDesktopSelectChange}
                onKeyDown={handleOptionKeydown}
                role="menuitemradio"
                aria-disabled={!!disabled}
                aria-checked={value === optionValue}
                tabIndex={isOpen && !disabled ? 0 : -1}
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...rest}
              >
                {label}
              </Style.DropdownOption>
            );
          })}
        </Style.DropdownOptionsContainer>
      </Style.DropdownContainer>
    </OffClickWrapper>
  );
};
