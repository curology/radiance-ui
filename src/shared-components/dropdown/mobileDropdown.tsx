import React from 'react';
import { useTheme } from '@emotion/react';

import { ChevronIcon } from '../../icons';
import Style from './style';

import { OptionType, OptionValue } from '.';

interface MobileDropdownProps<T> {
  borderRadius: string;
  id?: string;
  onMobileSelectChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  options: T[];
  textAlign: 'left' | 'center';
  value?: OptionValue;
}

/**
 * The mobile dropdown is built into the `<Dropdown />` component and cannot be used individually.
 *
 * `<Dropdown /> ` will determine if the user is on a mobile device and render a true `select` tag with `option`(s).
 */
export const MobileDropdown = <T extends OptionType>({
  borderRadius,
  id,
  onMobileSelectChange,
  options,
  textAlign,
  value,
}: MobileDropdownProps<T>) => {
  const theme = useTheme();

  return (
    <Style.DropdownContainer textAlign={textAlign}>
      <select
        css={Style.dropdownInputStyle({
          borderRadius,
          shouldBeFullyRounded: true,
          textAlign,
          theme,
        })}
        id={id}
        value={value ?? ''}
        onChange={onMobileSelectChange}
      >
        {options.map((option, index) => {
          let isDisabled = option.disabled;

          // Covers the case where default value is disabled
          // In mobile you cannot have a selected value as disabled option
          if (option.value === value) {
            isDisabled = false;
          }

          return (
            <option
              key={option.value ?? `undefined-${index}`}
              value={option.value}
              disabled={isDisabled}
            >
              {option.label}
            </option>
          );
        })}
      </select>
      <Style.IconContainer>
        <ChevronIcon width={10} height={10} />
      </Style.IconContainer>
    </Style.DropdownContainer>
  );
};
