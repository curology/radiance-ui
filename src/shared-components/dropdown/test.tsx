import React from 'react';
import { fireEvent, render, userEvent } from 'src/tests/testingLibraryHelpers';

import { DesktopDropdown } from './desktopDropdown';
import { MobileDropdown } from './mobileDropdown';

import { Dropdown } from './index';

const options = [
  { value: 'test1', label: 'Test1' },
  { value: 'test2', label: 'Test2' },
  { value: 'test3', label: 'Test3' },
];

const ON_CLICK_TEST = 'should be invoked onClick';

describe('<Dropdown />', () => {
  describe('on touch screen', () => {
    it('renders <MobileDropdown />', () => {
      window.document.documentElement.ontouchstart = () => undefined;
      const { getByRole, getAllByRole, queryAllByRole } = render(
        <Dropdown value="test1" options={options} onChange={() => undefined} />,
      );
      delete window.document.documentElement.ontouchstart;

      getByRole('combobox');
      expect(getAllByRole('option')).toHaveLength(3);
      expect(queryAllByRole('button')).toHaveLength(0);
    });
  });

  describe('when on non-touch screen', () => {
    it('renders <DesktopDropdown />', () => {
      const { getByRole } = render(
        <Dropdown value="test1" options={options} onChange={() => null} />,
      );

      // button only used in Desktop implementation
      expect(getByRole('button')).toBeTruthy();
    });
  });
});

describe('<MobileDropdown />', () => {
  describe('UI snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(
        <MobileDropdown
          onMobileSelectChange={() => undefined}
          borderRadius="4px"
          options={options}
          textAlign="left"
        />,
      );
      expect(container.firstElementChild).toMatchSnapshot();
    });
  });

  describe('onSelectChange callback', () => {
    it(`${ON_CLICK_TEST}`, () => {
      const spy = jest.fn();
      const { getByRole } = render(
        <MobileDropdown
          borderRadius="4px"
          options={options}
          onMobileSelectChange={spy}
          value=""
          textAlign="left"
        />,
      );

      const select = getByRole('combobox');
      fireEvent.change(select, { value: 'test1' });

      expect(spy).toHaveBeenCalled();
    });
  });
});

describe('<DesktopDropdown />', () => {
  it('renders the current option text', () => {
    const { getByRole } = render(
      <DesktopDropdown
        borderRadius="4px"
        closeDropdown={() => undefined}
        currentOption={{ value: 'test1', label: 'Test1' }}
        isOpen={false}
        onDesktopSelectChange={() => undefined}
        options={options}
        optionsContainerMaxHeight="250px"
        textAlign="left"
        toggleDropdown={() => undefined}
      />,
    );

    expect(getByRole('button', { name: 'Test1' })).toBeTruthy();
  });

  describe('onSelectClick callback', () => {
    it(`${ON_CLICK_TEST}`, () => {
      const spy = jest.fn();
      const { getByRole } = render(
        <DesktopDropdown
          borderRadius="4px"
          options={options}
          currentOption={{ value: 'test1', label: 'Test1' }}
          toggleDropdown={spy}
          optionsContainerMaxHeight="250px"
          closeDropdown={() => undefined}
          onDesktopSelectChange={() => undefined}
          textAlign="left"
          isOpen={false}
        />,
      );

      userEvent.click(getByRole('button'));
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('onOptionClick callback', () => {
    it(`${ON_CLICK_TEST}`, () => {
      const spy = jest.fn();
      const { getAllByRole } = render(
        <DesktopDropdown
          borderRadius="4px"
          options={options}
          currentOption={{ value: 'test1', label: 'Test1' }}
          onDesktopSelectChange={spy}
          isOpen
          optionsContainerMaxHeight="250px"
          toggleDropdown={() => null}
          closeDropdown={() => null}
          textAlign="left"
        />,
      );

      const listItems = getAllByRole('menuitemradio');
      // Arbitrarily select last item
      userEvent.click(listItems[listItems.length - 1]);
      expect(spy).toHaveBeenCalled();
    });
  });
});
