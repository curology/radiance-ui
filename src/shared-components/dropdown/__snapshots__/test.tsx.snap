// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileDropdown /> UI snapshots renders correctly 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
  text-align: left;
}

.emotion-0>div:first-of-type:focus {
  outline: none;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  box-shadow: 0px 8px 24px rgba(52, 51, 82, 0.05);
  background: #FFFFFF;
  background-image: none;
  width: 100%;
  min-height: 3.5rem;
  max-height: 3.5rem;
  border: 1px solid #EDEDF0;
  border-radius: 4px;
  color: #524D6E;
  line-height: 3.5rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 0 2rem 0 1rem;
  text-align: left;
  -webkit-text-align-last: left;
  text-align-last: left;
  -webkit-transition: 200ms ease-in-out;
  transition: 200ms ease-in-out;
  cursor: pointer;
}

.emotion-2:hover {
  box-shadow: 0px 8px 24px rgba(52, 51, 82, 0.10);
  -webkit-transition: 200ms ease-in-out;
  transition: 200ms ease-in-out;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0px 8px 24px rgba(52, 51, 82, 0.10);
  -webkit-transition: 200ms ease-in-out;
  transition: 200ms ease-in-out;
}

.emotion-2::-moz-focus-inner {
  border: 0;
}

.emotion-3 {
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  width: 1.5rem;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
}

.emotion-3:hover {
  cursor: pointer;
}

.emotion-3 svg {
  -webkit-transition: -webkit-transform 350ms ease-in-out;
  transition: transform 350ms ease-in-out;
}

.emotion-5 {
  display: block;
  -webkit-transform: rotate(0deg);
  -moz-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: color 350ms,-webkit-transform 350ms;
  transition: color 350ms,transform 350ms;
}

<div
  class="emotion-0 emotion-1"
>
  <select
    class="emotion-2"
  >
    <option
      value="test1"
    >
      Test1
    </option>
    <option
      value="test2"
    >
      Test2
    </option>
    <option
      value="test3"
    >
      Test3
    </option>
  </select>
  <div
    class="emotion-3 emotion-4"
  >
    <svg
      class="emotion-5"
      fill="none"
      height="10"
      viewBox="0 0 16 16"
      width="10"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.88 12.88L6 12l3.55-3.56L6 4.88 6.88 4l4.44 4.44-4.44 4.44z"
        fill="currentColor"
      />
    </svg>
  </div>
</div>
`;
